Maven build fails with Cycle detected when quarkus-openshift dependency is included

When both `quarkus-operator-sdk-csv-generator` and `quarkus-openshift` artifacts are included as dependencies for Maven build then `quarkus-maven-plugin` fails on `build` goal with `io.quarkus.builder.ChainBuildException: Cycle detected`.

Complete error:
```
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:2.4.1.Final:build (default-cli) on project test: Failed to build quarkus application: io.quarkus.builder.ChainBuildException: Cycle detected:
[ERROR] 		   io.quarkus.deployment.pkg.steps.JarResultBuildStep#buildRunnerJar produced class io.quarkus.deployment.pkg.builditem.JarBuildItem
[ERROR] 		to io.quarkus.container.image.openshift.deployment.OpenshiftProcessor#openshiftRequirementsJvm produced class io.quarkus.kubernetes.spi.DecoratorBuildItem
[ERROR] 		to io.quarkus.kubernetes.deployment.KubernetesProcessor#build produced class io.quarkus.kubernetes.spi.GeneratedKubernetesResourceBuildItem
[ERROR] 		to io.quarkiverse.operatorsdk.csv.deployment.CSVProcessor#generateCSV produced class io.quarkus.deployment.builditem.FeatureBuildItem
[ERROR] 		to io.quarkus.deployment.steps.MainClassBuildStep#build produced class io.quarkus.deployment.builditem.GeneratedClassBuildItem
[ERROR] 		to io.quarkus.deployment.pkg.steps.JarResultBuildStep#buildRunnerJar
```

Reproducer:
- Create pom.xml with this content:
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">

  <modelVersion>4.0.0</modelVersion>

  <groupId>com.test</groupId>
  <artifactId>test</artifactId>
  <version>0.0.1-SNAPSHOT</version>

  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>io.quarkus.platform</groupId>
        <artifactId>quarkus-bom</artifactId>
        <version>2.4.1.Final</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>

  <dependencies>
    <dependency>
      <groupId>io.quarkiverse.operatorsdk</groupId>
      <artifactId>quarkus-operator-sdk-csv-generator</artifactId>
      <version>2.0.0</version>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-openshift</artifactId>
    </dependency>
  </dependencies>

  <build>
    <plugins>
      <plugin>
        <groupId>io.quarkus.platform</groupId>
        <artifactId>quarkus-maven-plugin</artifactId>
        <extensions>true</extensions>
        <executions>
          <execution>
            <goals>
              <goal>build</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
    </plugins>
  </build>
</project>
```
- Run `mvn quarkus:build`

--- a/csv-generator/deployment/src/main/java/io/quarkiverse/operatorsdk/csv/deployment/CSVGenerator.java
+++ b/csv-generator/deployment/src/main/java/io/quarkiverse/operatorsdk/csv/deployment/CSVGenerator.java
@@ -1,9 +1,6 @@
 package io.quarkiverse.operatorsdk.csv.deployment;
 
-import java.io.File;
-import java.io.FileOutputStream;
 import java.io.IOException;
-import java.nio.file.Path;
 import java.util.Base64;
 import java.util.HashMap;
 import java.util.HashSet;
@@ -11,7 +8,9 @@ import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
 
 import org.jboss.logging.Logger;
 
@@ -27,7 +26,9 @@ import io.fabric8.kubernetes.api.model.rbac.PolicyRuleBuilder;
 import io.fabric8.kubernetes.api.model.rbac.Role;
 import io.fabric8.openshift.api.model.operatorhub.v1alpha1.ClusterServiceVersionBuilder;
 import io.fabric8.openshift.api.model.operatorhub.v1alpha1.ClusterServiceVersionFluent;
+import io.fabric8.openshift.api.model.operatorhub.v1alpha1.ClusterServiceVersionFluent.SpecNested;
 import io.fabric8.openshift.api.model.operatorhub.v1alpha1.ClusterServiceVersionSpecFluent;
+import io.fabric8.openshift.api.model.operatorhub.v1alpha1.ClusterServiceVersionSpecFluent.InstallNested;
 import io.fabric8.openshift.api.model.operatorhub.v1alpha1.NamedInstallStrategyFluent;
 import io.quarkiverse.operatorsdk.common.CustomResourceInfo;
 import io.quarkiverse.operatorsdk.csv.runtime.CSVMetadataHolder;
@@ -45,22 +46,69 @@ public class CSVGenerator {
         YAML_MAPPER.configure(SerializationFeature.WRITE_EMPTY_JSON_ARRAYS, false);
     }
 
-    public static void generate(Path outputDir, Map<String, AugmentedCustomResourceInfo> info,
-            Map<String, CSVMetadataHolder> csvMetadata,
-            String serviceAccountName, ClusterRole clusterRole, Role role, Deployment deployment) {
-        // load generated manifests
+    public static Set<NamedCSVBuilder> prepareGeneration(Map<String, AugmentedCustomResourceInfo> info,
+            Map<String, CSVMetadataHolder> csvMetadata) {
+        final var csvBuilders = new HashMap<String, NamedCSVBuilder>(7);
+        return info.values().parallelStream()
+                .map(cri -> csvBuilders.computeIfAbsent(cri.getCsvGroupName(),
+                        s -> new NamedCSVBuilder(cri, csvMetadata)))
+                .collect(Collectors.toSet());
+    }
 
-        final var controllerToCSVBuilders = new HashMap<String, ClusterServiceVersionBuilder>(7);
-        final var groupToCRInfo = new HashMap<String, Set<CustomResourceInfo>>(7);
-        info.forEach((crdName, cri) -> {
-            // record group name and associated CustomResourceInfos
+    private static Boolean hasMatchingClusterPermission(CustomResourceInfo cri,
+            NamedInstallStrategyFluent.SpecNested<ClusterServiceVersionSpecFluent.InstallNested<ClusterServiceVersionFluent.SpecNested<ClusterServiceVersionBuilder>>> installSpec,
+            Integer[] ruleIndex, Integer[] clusterPermissionIndex) {
+        final var hasMatchingClusterPermission = installSpec
+                .hasMatchingClusterPermission(cp -> {
+                    int i = 0;
+                    for (PolicyRule rule : cp.getRules()) {
+                        if (rule.getApiGroups().contains(cri.getGroup())) {
+                            ruleIndex[0] = i;
+                            return true;
+                        }
+                        i++;
+                    }
+                    clusterPermissionIndex[0]++;
+                    return false;
+                });
+        return hasMatchingClusterPermission;
+    }
+
+    static class NamedCSVBuilder {
+        private final String csvGroupName;
+        private final String controllerName;
+        private final ClusterServiceVersionBuilder csvBuilder;
+        final static Map<String, Set<CustomResourceInfo>> groupToCRInfo = new ConcurrentHashMap<>(7);
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            NamedCSVBuilder that = (NamedCSVBuilder) o;
+
+            return csvGroupName.equals(that.csvGroupName);
+        }
+
+        @Override
+        public int hashCode() {
+            return csvGroupName.hashCode();
+        }
+
+        public NamedCSVBuilder(AugmentedCustomResourceInfo cri, Map<String, CSVMetadataHolder> csvMetadata) {
+            // record group to CRI mapping
             groupToCRInfo.computeIfAbsent(cri.getGroup(), s -> new HashSet<>()).add(cri);
 
-            final var csvGroupName = cri.getCsvGroupName();
+            controllerName = cri.getControllerName();
+            csvGroupName = cri.getCsvGroupName();
             final var metadata = csvMetadata.get(csvGroupName);
-            final var csvSpecBuilder = controllerToCSVBuilders
-                    .computeIfAbsent(csvGroupName, s -> new ClusterServiceVersionBuilder()
-                            .withNewMetadata().withName(s).endMetadata())
+            csvBuilder = new ClusterServiceVersionBuilder()
+                    .withNewMetadata().withName(csvGroupName).endMetadata();
+            final var csvSpecBuilder = csvBuilder
                     .editOrNewSpec()
                     .withDescription(metadata.description)
                     .withDisplayName(metadata.displayName)
@@ -85,7 +133,7 @@ public class CSVGenerator {
             csvSpecBuilder
                     .editOrNewCustomresourcedefinitions()
                     .addNewOwned()
-                    .withName(crdName)
+                    .withName(cri.getCrdName())
                     .withVersion(cri.getVersion())
                     .withKind(cri.getKind())
                     .endOwned().endCustomresourcedefinitions();
@@ -94,7 +142,9 @@ public class CSVGenerator {
             final var installSpec = csvSpecBuilder.editOrNewInstall().editOrNewSpec();
             final Integer[] ruleIndex = new Integer[1];
             final Integer[] clusterPermissionIndex = new Integer[] { 0 };
-            final Boolean hasMatchingClusterPermission = hasMatchingClusterPermission(cri, installSpec, ruleIndex,
+            final Boolean hasMatchingClusterPermission = hasMatchingClusterPermission(
+                    cri,
+                    installSpec, ruleIndex,
                     clusterPermissionIndex);
             final var clusterPermission = hasMatchingClusterPermission
                     ? installSpec.editClusterPermission(clusterPermissionIndex[0])
@@ -117,128 +167,140 @@ public class CSVGenerator {
                 clusterPermission.setToRules(ruleIndex[0], rule.build());
             } else {
                 clusterPermission.addToRules(rule.addNewApiGroup(cri.getGroup())
-                        .addToVerbs("get", "list", "watch", "create", "delete", "patch", "update")
+                        .addToVerbs("get", "list", "watch", "create", "delete", "patch",
+                                "update")
                         .build());
             }
             clusterPermission.endClusterPermission();
             installSpec.endSpec().endInstall();
             csvSpecBuilder.endSpec();
-        });
+        }
 
-        controllerToCSVBuilders.forEach((controllerName, csvBuilder) -> {
-            final File file = new File(outputDir.toFile(), controllerName + ".csv.yml");
+        public String getFileName() {
+            return csvGroupName + ".csv.yml";
+        }
 
-            final var csvSpec = csvBuilder.editOrNewSpec();
+        private String getIconName() {
+            return csvGroupName + ".icon.png";
+        }
+
+        byte[] getYAMLData(String serviceAccountName, ClusterRole clusterRole, Role role,
+                Deployment deployment) throws IOException {
+            final var csvSpecBuilder = csvBuilder
+                    .editOrNewSpec();
+
             // deal with icon
             try (var iconAsStream = Thread.currentThread().getContextClassLoader()
-                    .getResourceAsStream(controllerName + ".icon.png");
-                    var outputStream = new FileOutputStream(file)) {
+                    .getResourceAsStream(getIconName())) {
                 if (iconAsStream != null) {
-                    final byte[] iconAsBase64 = Base64.getEncoder().encode(iconAsStream.readAllBytes());
-                    csvSpec.addNewIcon()
+                    final byte[] iconAsBase64 = Base64.getEncoder()
+                            .encode(iconAsStream.readAllBytes());
+                    csvSpecBuilder.addNewIcon()
                             .withBase64data(new String(iconAsBase64))
                             .withMediatype("image/png")
                             .endIcon();
                 }
+            } catch (IOException e) {
+                // ignore
+            }
 
-                final var installSpec = csvSpec.editOrNewInstall()
-                        .editOrNewSpec();
-                if (clusterRole != null) {
-                    // check if we have our CR group in the cluster role fragment and remove the one we added
-                    // before since we presume that if the user defined a fragment for permissions associated with their
-                    // CR they want that fragment to take precedence over automatically generated code
-                    final var rules = clusterRole.getRules();
-                    final var clusterPermissions = installSpec.buildClusterPermissions();
-                    groupToCRInfo.forEach((group, infos) -> {
+            final var installSpec = csvSpecBuilder.editOrNewInstall().editOrNewSpec();
+            handleClusterRole(serviceAccountName, clusterRole, groupToCRInfo, installSpec);
 
-                        final Predicate<PolicyRule> hasGroup = pr -> pr.getApiGroups().contains(group);
-                        final var nowEmptyPermissions = new LinkedList<Integer>();
-                        final var permissionPosition = new Integer[] { 0 };
-                        if (rules.stream().anyMatch(hasGroup)) {
-                            clusterPermissions.forEach(p -> {
-                                // record the position of all rules that match the group
-                                Integer[] index = new Integer[] { 0 };
-                                List<Integer> matchingRuleIndices = new LinkedList<>();
-                                p.getRules().forEach(r -> {
-                                    if (hasGroup.test(r)) {
-                                        matchingRuleIndices.add(index[0]);
-                                    }
-                                    index[0]++;
-                                });
+            handlerRole(serviceAccountName, role, installSpec);
 
-                                // remove the group from all matching rules
-                                matchingRuleIndices.forEach(i -> {
-                                    final var groups = p.getRules().get(i).getApiGroups();
-                                    groups.remove(group);
-                                    // if the rule doesn't have any groups anymore, remove it
-                                    if (groups.isEmpty()) {
-                                        p.getRules().remove(i.intValue());
-                                        // if the permission doesn't have any rules anymore, mark it for removal
-                                        if (p.getRules().isEmpty()) {
-                                            nowEmptyPermissions.add(permissionPosition[0]);
-                                        }
-                                    }
-                                });
+            handleDeployment(deployment, installSpec);
 
-                                permissionPosition[0]++;
+            // do not forget to end the elements!!
+            installSpec.endSpec().endInstall();
+            csvSpecBuilder.endSpec();
+
+            final var csv = csvBuilder.build();
+            return YAML_MAPPER.writeValueAsBytes(csv);
+
+        }
+
+        private void handleDeployment(Deployment deployment,
+                NamedInstallStrategyFluent.SpecNested<InstallNested<SpecNested<ClusterServiceVersionBuilder>>> installSpec) {
+            if (deployment != null) {
+                installSpec.addNewDeployment()
+                        .withName(deployment.getMetadata().getName())
+                        .withSpec(deployment.getSpec())
+                        .endDeployment();
+            }
+        }
+
+        private void handlerRole(String serviceAccountName, Role role,
+                NamedInstallStrategyFluent.SpecNested<InstallNested<SpecNested<ClusterServiceVersionBuilder>>> installSpec) {
+            if (role != null) {
+                installSpec
+                        .addNewPermission()
+                        .withServiceAccountName(serviceAccountName)
+                        .addAllToRules(role.getRules())
+                        .endPermission();
+            }
+        }
+
+        private void handleClusterRole(String serviceAccountName, ClusterRole clusterRole,
+                Map<String, Set<CustomResourceInfo>> groupToCRInfo,
+                NamedInstallStrategyFluent.SpecNested<InstallNested<SpecNested<ClusterServiceVersionBuilder>>> installSpec) {
+            if (clusterRole != null) {
+                // check if we have our CR group in the cluster role fragment and remove the one we added
+                // before since we presume that if the user defined a fragment for permissions associated with their
+                // CR they want that fragment to take precedence over automatically generated code
+                final var rules = clusterRole.getRules();
+                final var clusterPermissions = installSpec.buildClusterPermissions();
+                groupToCRInfo.forEach((group, infos) -> {
+
+                    final Predicate<PolicyRule> hasGroup = pr -> pr.getApiGroups()
+                            .contains(group);
+                    final var nowEmptyPermissions = new LinkedList<Integer>();
+                    final var permissionPosition = new Integer[] { 0 };
+                    if (rules.stream().anyMatch(hasGroup)) {
+                        clusterPermissions.forEach(p -> {
+                            // record the position of all rules that match the group
+                            Integer[] index = new Integer[] { 0 };
+                            List<Integer> matchingRuleIndices = new LinkedList<>();
+                            p.getRules().forEach(r -> {
+                                if (hasGroup.test(r)) {
+                                    matchingRuleIndices.add(index[0]);
+                                }
+                                index[0]++;
                             });
 
-                            // remove now empty permissions
-                            nowEmptyPermissions.forEach(i -> clusterPermissions.remove(i.intValue()));
-                            installSpec.addAllToClusterPermissions(clusterPermissions);
-                        }
-                    });
-                    installSpec
-                            .addNewClusterPermission()
-                            .withServiceAccountName(serviceAccountName)
-                            .addAllToRules(rules)
-                            .endClusterPermission();
-                }
+                            // remove the group from all matching rules
+                            matchingRuleIndices.forEach(i -> {
+                                final var groups = p.getRules().get(i).getApiGroups();
+                                groups.remove(group);
+                                // if the rule doesn't have any groups anymore, remove it
+                                if (groups.isEmpty()) {
+                                    p.getRules().remove(i.intValue());
+                                    // if the permission doesn't have any rules anymore, mark it for removal
+                                    if (p.getRules().isEmpty()) {
+                                        nowEmptyPermissions.add(permissionPosition[0]);
+                                    }
+                                }
+                            });
 
-                if (role != null) {
-                    installSpec
-                            .addNewPermission()
-                            .withServiceAccountName(serviceAccountName)
-                            .addAllToRules(role.getRules())
-                            .endPermission();
-                }
+                            permissionPosition[0]++;
+                        });
 
-                if (deployment != null) {
-                    installSpec.addNewDeployment()
-                            .withName(deployment.getMetadata().getName())
-                            .withSpec(deployment.getSpec())
-                            .endDeployment();
-                }
-
-                // do not forget to end the elements!!
-                installSpec.endSpec().endInstall();
-                csvSpec.endSpec();
-
-                final var csv = csvBuilder.build();
-                YAML_MAPPER.writeValue(outputStream, csv);
-                log.infov("Generated CSV for {0} controller -> {1}", controllerName, file);
-            } catch (IOException e) {
-                e.printStackTrace();
-            }
-        });
-    }
-
-    private static Boolean hasMatchingClusterPermission(CustomResourceInfo cri,
-            NamedInstallStrategyFluent.SpecNested<ClusterServiceVersionSpecFluent.InstallNested<ClusterServiceVersionFluent.SpecNested<ClusterServiceVersionBuilder>>> installSpec,
-            Integer[] ruleIndex, Integer[] clusterPermissionIndex) {
-        final var hasMatchingClusterPermission = installSpec
-                .hasMatchingClusterPermission(cp -> {
-                    int i = 0;
-                    for (PolicyRule rule : cp.getRules()) {
-                        if (rule.getApiGroups().contains(cri.getGroup())) {
-                            ruleIndex[0] = i;
-                            return true;
-                        }
-                        i++;
+                        // remove now empty permissions
+                        nowEmptyPermissions.forEach(
+                                i -> clusterPermissions.remove(i.intValue()));
+                        installSpec.addAllToClusterPermissions(clusterPermissions);
                     }
-                    clusterPermissionIndex[0]++;
-                    return false;
                 });
-        return hasMatchingClusterPermission;
+                installSpec
+                        .addNewClusterPermission()
+                        .withServiceAccountName(serviceAccountName)
+                        .addAllToRules(rules)
+                        .endClusterPermission();
+            }
+        }
+
+        public String getControllerName() {
+            return controllerName;
+        }
     }
 }

--- a/csv-generator/deployment/src/main/java/io/quarkiverse/operatorsdk/csv/deployment/CSVProcessor.java
+++ b/csv-generator/deployment/src/main/java/io/quarkiverse/operatorsdk/csv/deployment/CSVProcessor.java
@@ -3,6 +3,8 @@ package io.quarkiverse.operatorsdk.csv.deployment;
 import static io.quarkus.kubernetes.deployment.Constants.KUBERNETES;
 
 import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.nio.file.Path;
 import java.util.HashMap;
 import java.util.List;
 
@@ -28,7 +30,7 @@ import io.quarkiverse.operatorsdk.deployment.GeneratedCRDInfoBuildItem;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
-import io.quarkus.deployment.pkg.builditem.ArtifactResultBuildItem;
+import io.quarkus.deployment.builditem.GeneratedFileSystemResourceBuildItem;
 import io.quarkus.deployment.pkg.builditem.OutputTargetBuildItem;
 import io.quarkus.kubernetes.spi.GeneratedKubernetesResourceBuildItem;
 
@@ -66,10 +68,9 @@ public class CSVProcessor {
     @BuildStep
     void generateCSV(OutputTargetBuildItem outputTarget,
             CSVMetadataBuildItem csvMetadata,
-            BuildProducer<GeneratedCSVBuildItem> generatedCSV,
+            BuildProducer<GeneratedCSVBuildItem> doneGeneratingCSV,
             List<GeneratedKubernetesResourceBuildItem> generatedKubernetesManifests,
-            // this is added to ensure that the build step will be run
-            BuildProducer<ArtifactResultBuildItem> forceExtensionToRun) {
+            BuildProducer<GeneratedFileSystemResourceBuildItem> generatedCSVs) {
         if (configuration.generateCSV.orElse(false)) {
             try {
                 final var outputDir = outputTarget.getOutputDirectory().resolve(KUBERNETES);
@@ -107,9 +108,23 @@ public class CSVProcessor {
                                         }
                                     });
                                 });
-                CSVGenerator.generate(outputDir, csvMetadata.getAugmentedCustomResourceInfos(), csvMetadata.getCSVMetadata(),
-                        serviceAccountName[0], clusterRole[0], role[0], deployment[0]);
-                generatedCSV.produce(new GeneratedCSVBuildItem());
+                final var generated = CSVGenerator.prepareGeneration(
+                        csvMetadata.getAugmentedCustomResourceInfos(), csvMetadata.getCSVMetadata());
+                generated.forEach(namedCSVBuilder -> {
+                    final var fileName = namedCSVBuilder.getFileName();
+                    try {
+                        generatedCSVs.produce(
+                                new GeneratedFileSystemResourceBuildItem(
+                                        Path.of(KUBERNETES, fileName).toString(),
+                                        namedCSVBuilder.getYAMLData(serviceAccountName[0],
+                                                clusterRole[0], role[0], deployment[0])));
+                        log.infov("Generating CSV for {0} controller -> {1}", namedCSVBuilder.getControllerName(),
+                                outputDir.resolve(fileName));
+                    } catch (IOException e) {
+                        log.errorv("Cannot generate CSV for {0}: {1}", namedCSVBuilder.getControllerName(), e.getMessage());
+                    }
+                });
+                doneGeneratingCSV.produce(new GeneratedCSVBuildItem());
             } catch (Exception e) {
                 log.infov(e, "Couldn't generate CSV:");
             }

--- a/csv-generator/deployment/src/main/java/io/quarkiverse/operatorsdk/csv/deployment/CSVProcessor.java
+++ b/csv-generator/deployment/src/main/java/io/quarkiverse/operatorsdk/csv/deployment/CSVProcessor.java
@@ -28,7 +28,7 @@ import io.quarkiverse.operatorsdk.deployment.GeneratedCRDInfoBuildItem;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
-import io.quarkus.deployment.builditem.FeatureBuildItem;
+import io.quarkus.deployment.pkg.builditem.ArtifactResultBuildItem;
 import io.quarkus.deployment.pkg.builditem.OutputTargetBuildItem;
 import io.quarkus.kubernetes.spi.GeneratedKubernetesResourceBuildItem;
 
@@ -64,10 +64,12 @@ public class CSVProcessor {
     }
 
     @BuildStep
-    FeatureBuildItem generateCSV(OutputTargetBuildItem outputTarget,
+    void generateCSV(OutputTargetBuildItem outputTarget,
             CSVMetadataBuildItem csvMetadata,
-            BuildProducer<GeneratedCSVBuildItem> ignored,
-            List<GeneratedKubernetesResourceBuildItem> generatedKubernetesManifests) {
+            BuildProducer<GeneratedCSVBuildItem> generatedCSV,
+            List<GeneratedKubernetesResourceBuildItem> generatedKubernetesManifests,
+            // this is added to ensure that the build step will be run
+            BuildProducer<ArtifactResultBuildItem> forceExtensionToRun) {
         if (configuration.generateCSV.orElse(false)) {
             try {
                 final var outputDir = outputTarget.getOutputDirectory().resolve(KUBERNETES);
@@ -107,14 +109,11 @@ public class CSVProcessor {
                                 });
                 CSVGenerator.generate(outputDir, csvMetadata.getAugmentedCustomResourceInfos(), csvMetadata.getCSVMetadata(),
                         serviceAccountName[0], clusterRole[0], role[0], deployment[0]);
-                ignored.produce(new GeneratedCSVBuildItem());
+                generatedCSV.produce(new GeneratedCSVBuildItem());
             } catch (Exception e) {
                 log.infov(e, "Couldn't generate CSV:");
             }
         }
-
-        // generating a feature is a sure way to make sure this step will be executed by Quarkus
-        return new FeatureBuildItem("CSVGeneration");
     }
 
     private CSVMetadataHolder getCSVMetadata(ClassInfo info, String controllerName, IndexView index) {

--- a/samples/pom.xml
+++ b/samples/pom.xml
@@ -38,6 +38,10 @@
       <groupId>io.quarkus</groupId>
       <artifactId>quarkus-rest-client-jackson</artifactId>
     </dependency>
+    <dependency>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-openshift</artifactId>
+    </dependency>
     <!--  This dependency is needed only to ensure proper building order so that this module is build after the CSV extension -->
     <dependency>
       <groupId>io.quarkiverse.operatorsdk</groupId>

--- a/samples/src/main/resources/application.properties
+++ b/samples/src/main/resources/application.properties
@@ -2,3 +2,4 @@ joke-api/mp-rest/url=https://v2.jokeapi.dev/joke
 joke-api/mp-rest/scope=javax.inject.Singleton
 quarkus.operator-sdk.crd.apply=true
 quarkus.operator-sdk.generate-csv=true
+quarkus.container-image.builder=jib

